src/libopensc/card-openpgp.c

```diff
@@ -2082,6 +2175,152 @@ pgp_put_data(sc_card_t *card, unsigned int tag, const u8 *buf, size_t buf_len)
 	LOG_FUNC_RETURN(card->ctx, (int)buf_len);
 }
 
+#ifdef ENABLE_OPENSSL
+/* KDF_ITERSALTED_S2K algorithm, a simplified version of openpgp_s2k from libgcrypt[1] that
+ * has only only one-pass and its output key size same as the size of hash algorithm specified.
+ * [1]: https://github.com/gpg/libgcrypt/blob/libgcrypt-1.11.0/cipher/kdf.c#L32-L109 */
+static int
+kdf_itersalted_s2k(struct sc_context *ctx, const char *hash, const uint8_t *pin, size_t pinlen, const uint8_t *salt, size_t saltlen, uint32_t iterations, uint8_t **out, size_t *outlen)
+{
+	if (!ctx || !hash || !pin || pinlen == 0 || !salt || saltlen != 8 || !out || !outlen) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	EVP_MD *md = sc_evp_md(ctx, hash);
+	if (md == NULL) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	EVP_MD_CTX *hctx = EVP_MD_CTX_new();
+	if (hctx == NULL) {
+		sc_evp_md_free(md);
+		return SC_ERROR_INTERNAL;
+	}
+	if (!EVP_DigestInit(hctx, md)) {
+		EVP_MD_CTX_free(hctx);
+		sc_evp_md_free(md);
+		return SC_ERROR_INTERNAL;
+	}
+	int keysize = EVP_MD_size(md);
+	if (keysize <= 0) {
+		EVP_MD_CTX_free(hctx);
+		sc_evp_md_free(md);
+		return SC_ERROR_INTERNAL;
+	}
+	uint8_t *outkey = (uint8_t *)calloc(1, keysize);
+	if (outkey == NULL) {
+		EVP_MD_CTX_free(hctx);
+		sc_evp_md_free(md);
+		return SC_ERROR_NOT_ENOUGH_MEMORY;
+	}
+	size_t count = iterations;
+	if (count < pinlen + saltlen) {
+		count = pinlen + saltlen;
+	}
+	while (count >= pinlen + saltlen) {
+		EVP_DigestUpdate(hctx, salt, saltlen);
+		EVP_DigestUpdate(hctx, pin, pinlen);
+		count -= pinlen + saltlen;
+	}
+	if (count <= saltlen) {
+		EVP_DigestUpdate(hctx, salt, count);
+	} else {
+		EVP_DigestUpdate(hctx, salt, saltlen);
+		count -= saltlen;
+		EVP_DigestUpdate(hctx, pin, count);
+	}
+
+	unsigned int outsize = keysize;
+	EVP_DigestFinal(hctx, outkey, &outsize);
+	EVP_MD_CTX_free(hctx);
+	sc_evp_md_free(md);
+	if (outsize != (unsigned int)keysize) {
+		free(outkey);
+		return SC_ERROR_INTERNAL;
+	}
+	*out = outkey;
+	*outlen = outsize;
+	return SC_SUCCESS;
+}
+
+static int
+pgp_kdf_do_pin_cmd(sc_card_t *card, struct sc_pin_cmd_data *data, int *tries_left)
+{
+	struct pgp_priv_data *priv = DRVDATA(card);
+	switch (data->cmd) {
+	case SC_PIN_CMD_VERIFY:
+	case SC_PIN_CMD_CHANGE:
+	case SC_PIN_CMD_UNBLOCK:
+		break;
+	default:
+		LOG_FUNC_RETURN(card->ctx, iso_ops->pin_cmd(card, data, tries_left));
+	}
+	pgp_pin_kdf_info *info = priv->pin_kdf_info;
+	uint8_t *pin1_derived = NULL;
+	uint8_t *pin2_derived = NULL;
+	size_t pin1_derived_size = 0;
+	size_t pin2_derived_size = 0;
+	const uint8_t *pin1 = data->pin1.data;
+	size_t pin1len = data->pin1.len;
+	const uint8_t *pin2 = data->pin2.data;
+	size_t pin2len = data->pin2.len;
+	uint8_t *salt = NULL;
+	size_t saltlen = 0;
+	switch (data->pin_reference) {
+	case 0x81:
+	case 0x82:
+		salt = info->userpw_salt;
```

Comment:
need to check if `info != NULL`

Reply:
It's already checked by the caller.

Reply:
That's incorrect. if the initial calloc fails, then this is not fatal to the initialization of the openpgp driver and the priv's member will be NULL. please be generous with sanity checks - if they are not needed then they are either optimized out by the compiler or they are extremely cheap given modern CPUs.

Reply:
`pgp_kdf_do_pin_cmd()` will only be called if `priv->pin_kdf_info` is not NULL. But I'll add a check anyway.